src/backend/distributed/planner/multi_planner_utils.c

```diff
@@ -0,0 +1,863 @@
+/*
+ * multi_planner_utils.c
```

Comment:
I still think shard_pruning.c would be better (we've started to move away from multi_ prefixes btw), but I can live with that.